render-ws-spark-client/src/main/java/org/janelia/render/client/spark/intensityadjust/ShadingCorrectionTileClient.java

```
@@ -0,0 +1,245 @@
+        final Broadcast<Bounds> stackBoundsBroadcast = sparkContext.broadcast(stackBounds);
+        final Broadcast<List<Double>> zValuesBroadcast = sparkContext.broadcast(zValues);
+
+        final List<Integer> zIndices = IntStream.range(0, zValues.size()).boxed().collect(Collectors.toList());
+        sparkContext.parallelize(zIndices)
+                .foreach(zIndex -> {
+                    final Double z = zValuesBroadcast.getValue().get(zIndex);
+                    final ShadingModel layerModel = modelProviderBroadcast.getValue().getModel(z.intValue());
+                    addShadingCorrectionToLayer(z, layerModel, parametersBroadcast.value(), stackBoundsBroadcast.value());
+                });
+
+        completeTargetStack(renderClient);
+
+        LOG.info("runWithContext: exit");
+    }
+
+    private void setUpTargetStack(final RenderDataClient dataClient) throws IOException {
+        final StackMetaData sourceStackMetaData = dataClient.getStackMetaData(parameters.stack);
+        dataClient.setupDerivedStack(sourceStackMetaData, parameters.targetStack);
+        LOG.info("setUpTargetStack: setup stack {}", parameters.targetStack);
+    }
+
+    private void completeTargetStack(final RenderDataClient dataClient) throws IOException {
+        dataClient.setStackState(parameters.targetStack, StackMetaData.StackState.COMPLETE);
+        LOG.info("completeTargetStack: setup stack {}", parameters.targetStack);
+    }
+
+    private void addShadingCorrectionToLayer(
+            final Double z,
+            final ShadingModel layerModel,
+            final Parameters parameters,
+            final Bounds bounds
+    ) throws IOException {
+
+        final RenderDataClient renderClient = parameters.webservice.getDataClient();
+        final ResolvedTileSpecCollection rtsc = renderClient.getResolvedTiles(parameters.stack, z);
+
+        if (layerModel == null) {
+            LOG.warn("No model found for z={}", z);
```

If null layer models are expected then this is fine but if it really is an error case I'd throw an exception here.

---

They were kind of expected as a way to skip shading correction on certain layers. I made this logic explicit and a little more robust.